parsers/ocaml.c

```diff
@@ -1864,25 +2067,55 @@ static void findOcamlTags (void)
 	lexingState st;
 	ocaToken tok;
 
+	/* One-token lookahead gives us the ability to
+	 * do much more accurate analysis */
+	lexingState nextSt;
+	ocaToken nextTok;
+
 	initStack ();
+
 	computeModuleName ();
 	tempIdent = vStringNew ();
 	lastModule = vStringNew ();
 	lastClass = vStringNew ();
 	voidName = vStringNew ();
 	vStringCopyS (voidName, "_");
+	vString *temp_cp = vStringNew ();
 
-	st.name = vStringNew ();
-	st.cp = fileReadLine ();
+	nextSt.name = vStringNew ();
+	nextSt.cp = fileReadLine ();
 	toDoNext = &globalScope;
-	tok = lex (&st);
+	nextTok = lex (&nextSt);
+
+	/* prime the lookahead token */
+	st = nextSt;	// preserve the old state for our first token
+	st.name = vStringNewCopy (st.name);
+	st.cp = (const unsigned char *) temp_cp->buffer;
```

Comment:
well, here you copy `nextSt` (which contains the *first* token at this point) to `st`, so I would think you should also copy the `cp` `lex(&nextSt)` above maybe have moved.

Assuming you just want to copy the `nextSt` already read (so you read it again (below) to get the actual next (second) token), I can't imagine you don't want to keep the whole original (first token) `nextSt`, including `cp`.

BTW, IMO it'd be less confusing if you'd first read in `st` and *then* in `nextSt`, instead or reading twice in `nextSt` and copying in `st`:
```diff
diff --git a/parsers/ocaml.c b/parsers/ocaml.c
index f321fd1..a9baed8 100644
--- a/parsers/ocaml.c
+++ b/parsers/ocaml.c
@@ -2082,19 +2082,21 @@ static void findOcamlTags (void)
     vStringCopyS (voidName, "_");
     vString *temp_cp = vStringNew ();
 
-    nextSt.name = vStringNew ();
-    nextSt.cp = fileReadLine ();
+    st.name = vStringNew ();
+    st.cp = fileReadLine ();
     toDoNext = &globalScope;
-    nextTok = lex (&nextSt);
+    tok = lex (&st);
 
-    if (nextTok != Tok_EOF)
+    if (tok != Tok_EOF)
         computeModuleName ();
 
     /* prime the lookahead token */
-    st = nextSt;    // preserve the old state for our first token
-    st.name = vStringNewCopy (st.name);
+    nextSt.name = vStringNewCopy (st.name);
+    /* backup previous cp and swap it, to have the real one (fileReadLine()
+     * result) in the token we actually read */
+    vStringCopyS (temp_cp, (const char *) nextSt.cp);
+    nextSt.cp = st.cp;
     st.cp = (const unsigned char *) temp_cp->buffer;
-    tok = nextTok;
     ocaLineNumber = File.source.lineNumber;
     ocaFilePosition = File.filePosition;
     nextTok = lex (&nextSt);
```
or am I missing something?